src/libsyntax/diagnostics/metadata.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+/// This module contains utilities for outputting extended-error metadata.
+
+use std::collections::BTreeMap;
+use std::path::{Path, PathBuf};
+use std::fs::{read_dir, create_dir_all, OpenOptions, File};
+use std::io::{Read, Write};
+use std::error::Error;
+use rustc_serialize::json::{self, as_json};
+
+use codemap::Span;
+use ext::base::ExtCtxt;
+use diagnostics::plugin::{ErrorMap, ErrorInfo};
+
+pub use self::Uniqueness::*;
+
+// Directory to use for extended error JSON.
+const ERROR_METADATA_DIR: &'static str = "tmp/extended-errors";
+
+/// JSON encodable/decodable version of `ErrorInfo`.
+#[derive(PartialEq, RustcDecodable, RustcEncodable)]
+struct ErrorMetadata {
+    pub description: Option<String>,
+    pub use_site: Option<ErrorLocation>
+}
+
+type ErrorMetadataMap = BTreeMap<String, ErrorMetadata>;
+
+/// JSON encodable error location type with filename and line number.
+#[derive(PartialEq, RustcDecodable, RustcEncodable)]
+struct ErrorLocation {
+    pub filename: String,
+    pub line: usize
+}
+
+impl ErrorLocation {
+    /// Create an error location from a span.
+    fn from_span(ecx: &ExtCtxt, sp: Span) -> ErrorLocation {
+        let loc = ecx.codemap().lookup_char_pos_adj(sp.lo);
+        ErrorLocation {
+            filename: loc.filename,
+            line: loc.line
+        }
+    }
+}
+
+/// Type for describing the uniqueness of a set of error codes, as returned by `check_uniqueness`.
+pub enum Uniqueness {
+    /// All errors in the crate checked are unique (so far).
+    Unique,
+    /// Info on the first duplicate, of the form:
+    /// (error-code, metadata filename where duplicate appears)
+    Duplicate(String, String)
+}
+
+
+/// Get the path where error metadata for a given crate should be stored.
+pub fn get_metadata_path(crate_name: &str) -> PathBuf {
+    Path::new(ERROR_METADATA_DIR).join(format!("{}.json", crate_name))
+}
+
+/// Check that the errors used in a given crate are unique across the compiler.
+/// This is achieved by reading metadata files for other crates and checking them.
```

Comment:
Nit: these comments would be better phrased in terms of the actual inputs to the functions, rather than there proposed uses. I.e., something like, "Checks that no errors defined in err_map are duplicated in any of the error metadata files in the directory given by crate_file". Since I don't think there is any requirement for anything here to be tied to crates.